diff --git a/nptl/pthread_create.c b/nptl/pthread_create.c
index 5682c9c2c0..e2f3c7cb30 100644
--- a/nptl/pthread_create.c
+++ b/nptl/pthread_create.c
@@ -20,6 +20,8 @@
 #include <errno.h>
 #include <stdbool.h>
 #include <stdlib.h>
+#include <stdio.h>
+#include <time.h>
 #include <string.h>
 #include <stdint.h>
 #include "pthreadP.h"
@@ -619,13 +621,71 @@ report_thread_creation (struct pthread *pd)
 }
 
 
+int __custom_scheduling (pthread_attr_t *attr)
+{
+  int pthread_ret = 0;
+
+  // configure priority (default: static value 90)
+  srandom(time(NULL));
+  const char *rand_prio = getenv("RAND_PRIO");
+  int prio = (rand_prio != NULL && atoi(rand_prio) == 1)? random()%100: 90;
+  const struct sched_param param = { prio };
+
+  // configure sched policy (default: SCHED_FIFO)
+  const char *sched_policy_str = getenv("SCHED_POLICY");
+  int sched_policy = (sched_policy_str == NULL)? SCHED_FIFO: atoi(sched_policy_str);
+
+  cpu_set_t cpuset;
+  CPU_ZERO(&cpuset);
+  CPU_SET(0, &cpuset);
+
+  if ((pthread_ret = pthread_attr_init(attr)) != 0)
+  {
+    printf("Error calling pthread_attr_init: %s\n", strerror(pthread_ret));
+  }
+  else if ((pthread_ret = pthread_attr_setschedpolicy(attr, sched_policy)) != 0)
+  {
+    printf("Error calling pthread_attr_setschedpolicy: %s\n", strerror(pthread_ret));
+  }
+  else if ((pthread_ret = pthread_attr_setschedparam(attr, &param)) != 0)
+  {
+    printf("Error calling pthread_attr_setschedparam: %s\n", strerror(pthread_ret));
+  }
+  else if ((pthread_ret = pthread_attr_setinheritsched(attr, PTHREAD_EXPLICIT_SCHED)) != 0)
+  {
+    printf("Error calling pthread_attr_setinheritsched: %s\n", strerror(pthread_ret));
+  }
+  else if ((pthread_ret = pthread_attr_setaffinity_np(attr, sizeof(cpu_set_t), &cpuset)) != 0)
+  {
+    printf("Error calling pthread_attr_setaffinity_np: %s\n", strerror(pthread_ret));
+  }
+  return pthread_ret;
+}
+
 int
 __pthread_create_2_1 (pthread_t *newthread, const pthread_attr_t *attr,
 		      void *(*start_routine) (void *), void *arg)
 {
   STACK_VARIABLES;
 
-  const struct pthread_attr *iattr = (struct pthread_attr *) attr;
+  pthread_attr_t inject_attr;
+
+  const char *enable_custom_sched = getenv("CUSTOM_SCHED");
+  int custom_sched_success = 0;
+  if (enable_custom_sched != NULL && strncmp(enable_custom_sched, "1", 1) == 0)
+  {
+    if (__custom_scheduling(&inject_attr) != 0)
+    {
+      printf("Error assigning random priority to thread\n");
+    }
+    else
+    {
+      custom_sched_success = 1;
+    }
+  }
+  const struct pthread_attr *iattr = custom_sched_success?
+    (struct pthread_attr *) &inject_attr: (struct pthread_attr *) attr;
+
   struct pthread_attr default_attr;
   bool free_cpuset = false;
   bool c11 = (attr == ATTR_C11_THREAD);
